extensions/csrc/kernel/cuda/fused_rotary_emb_and_cache_kernel.cu

```diff
@@ -220,14 +225,39 @@ __global__ void rotary_embedding_kernel(
     apply_emb_rotary_compute<scalar_t, m_scalar_t, VecSize>(key, cos_ptr, sin_ptr, key_stride, token_id, shard_block_size, half_head_dim, kv_head_num, head_dim);
```

Comment:
You can abstract the logic of k rotary embedding and kv memcopy, and implement them as separate device sub-functions. Then, incorporate the separated k rotary embedding device sub-function into the rotary_embedding_kernel function to achieve layout adaptability for the rotary_embedding_kernel.

Reply:
As the fused_rotary_emb_and_cache kernel utilize a different thread mapping strategy (according to rotary_embedding) when compared with kv_memcpy, you cannot directly call kv_memcpy for k cache